pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,434 @@
+                        continue;
+                    }
+                };
+                e.insert(tenant)
+            }
+        };
+        let timeline = match held_timelines.entry(timeline_id) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => {
+                let timeline = match tenant.get_timeline(*timeline_id, false) {
+                    Ok(timeline) => timeline,
+                    Err(e) => {
+                        warn! {%tenant_id, %timeline_id, "failed to get timeline: {:#}", e};
+                        continue;
+                    }
+                };
+                e.insert(timeline)
+            }
+        };
+        // XXX batching? but, if the batch fails due to one layer, we should retry the remaining ones
+        match timeline.evict_layer(&layer.file_name.file_name()).await {
+            Err(e) => {
+                warn! { %tenant_id, %timeline_id, "failed to evict layer: {:#}", e};
+            }
+            Ok(None) | Ok(Some(false)) => {
+                info! { %tenant_id, %timeline_id, "layer does not exist, likely it was deleted" };
+                // XXX should we account for this in usage_expected.avail_bytes ?
+            }
+            Ok(Some(true)) => {
+                usage_expected.avail_bytes += layer.file_size;
+            }
+        }
+        if !usage_expected.with_pressure(task_config).has_pressure() {
+            break;
+        }
+    }
+
+    // Verify with statvfs whether we made any real progress
+    let usage_post = get_filesystem_level_usage(tenants_dir_fd)
+        // It's quite unlikely to hit the error here. Keep the code simple and bail out.
+        .context("get filesystem-level disk usage after evictions")?
+        .with_pressure(task_config);
+
+    debug!(?usage_post, "disk usage");
+
+    Ok(IterationOutcome::Finished(IterationOutcomeFinished {
+        before: usage_pre,
+        after_actual: usage_post,
+        after_expected: usage_expected,
+    }))
+}
+
+#[derive(Debug, Clone)]
+// The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+// We use the Debug impl for logging, so, it's allright.
+#[allow(dead_code)]
+
+struct FilesystemLevelDiskUsage {
+    pub total_bytes: u64,
+    pub avail_bytes: u64,
+    pub usage_pct: u64,
+}
+
+type FilesystemLevelDiskUsagePressure = [(&'static str, bool); 2];
+
+#[derive(Debug)]
+// The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+// We use the Debug impl for logging, so, it's allright.
+#[allow(dead_code)]
+
+struct FilesystemLevelDiskUsageWithPressure {
+    pub usage: FilesystemLevelDiskUsage,
+    pub pressure: FilesystemLevelDiskUsagePressure,
+}
+
+impl FilesystemLevelDiskUsageWithPressure {
+    #[inline]
+    fn has_pressure(&self) -> bool {
+        self.pressure.iter().any(|(_, has_pressure)| *has_pressure)
+    }
+}
+impl FilesystemLevelDiskUsage {
+    #[inline]
```

Comment:
These are meaningless within one crate. More generally, with LTO, all functions are `#[inline]`, `#[inline(always)]` is the hint which should be used sometimes, but I doubt you measured any perf difference here? :)